flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBase.java

```diff
@@ -494,6 +514,12 @@ public void open(Configuration configuration) throws Exception {
 				}
 			}
 
+			if (filterRestoredPartitionsWithDiscovered) {
+				subscribedPartitionsToStartOffsets.entrySet().removeIf(
+					entry -> (!allPartitions.contains(entry.getKey()))
```

Comment:
I'm not entirely sure this is safe.
This is relying on the fact that Kafka indeed returns the correct set of partitions based on the `topicsDescriptor`.
While in normal conditions, the end results should be the same, but I feel like it is more safer to filter based on the `topicsDescriptor` so we're not relying on external system behaviours and subject to any system hiccups.

@fengli What do you think?

Reply:
Could you also add warning logs that the offset states are filtered out?
We should make this visible to some degree, since we're essentially dropping state.

Reply:
Thanks for the review @tzulitai , Appreciate it.

1. 
- You are talking about the topicDescriptor will be different from what we will discovered with Kafka? TopicDescriptor is user inputs, can be invalid topics. I think it might be better to rely on what Kafka tells us (And it's KafkaConsumerBase, relies heavily on Kafka). 
- If we rely on topicDescriptor, we need to check against fixedtopics and pattertopics. Will probably add a bit more complexity here.

What do you think? But I'm OK to use topicDescriptor if you think it's a good direction to go.

2. Done

Reply:
@fengli 
Well, if it is invalid topics, then the `allPartitions` wouldn't contain the correct partitions anyways.

I still think that using `topicsDescriptor` is the safer way to do this.
For example, lets say that due to some Kafka system hiccup, Kafka temporarily did not return some partition for a given topic. If we used `allPartitions` to filter out the restored offset states, that partition will be filtered out.
I'm not sure this will ever happen, but again, it's something out of the connector code's control. And using such information to drop state is something I feel uneasy about.

Reply:
Yes, if we use the `topicsDescriptor`, it'll likely require a bit more code.

Perhaps we can add a method:
`boolean KafkaTopicsDescriptor#isMatchingTopic(String)`, which essentially just does:
```
if (isFixedTopics()) {
    return fixedTopics.contain(topic)
} else {
    return topicPattern.matcher(topic).matches()
}
```

Can maybe even use this new method to simplify `AbstractPartitionDiscoverer` a bit.

Either way, this would be a preliminary code refactoring. For this kind of changes, I suggest having a separate commit in the PR before the actual change.

Reply:
Done.